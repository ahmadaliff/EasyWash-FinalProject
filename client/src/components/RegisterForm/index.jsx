import PropTypes from 'prop-types';
import { useState } from 'react';
import toast from 'react-hot-toast';
import { useForm } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import PhoneInput from 'react-phone-number-input';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage, injectIntl } from 'react-intl';

import { FormControl } from '@mui/material';

import InputRHF from '@components/InputRHF';

import intlHelper from '@utils/intlHelper';

import { actionSetStep } from '@pages/Register/actions';
import { actionHandleRegister } from '@containers/Client/actions';
import { selectEmail, selectRole, selectStep } from '@pages/Register/selectors';

import classes from '@components/RegisterForm/style.module.scss';
import 'react-phone-number-input/style.css';

const RegisterForm = ({ step, role, email, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [showPass, setShowPass] = useState(false);
  const [phone, setPhone] = useState(null);

  const {
    handleSubmit,
    register,
    formState: { errors },
    watch,
  } = useForm();
  const fullName = watch('fullName');
  const password = watch('password');

  const onSubmit = (data) => {
    if (data.password !== data.confirmPassword) {
      toast.error(intlHelper({ message: 'app_pass_not_same' }));
    } else {
      data.email = email;
      data.role = role;
      data.phone = phone;
      delete data.confirmPassword;
      dispatch(
        actionHandleRegister(data, () => {
          setTimeout(() => {
            navigate('/login');
          }, 1000);
        })
      );
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} data-testid="form-register">
      <InputRHF
        input={{
          name: 'fullName',
          required: formatMessage({ id: 'app_user_fullName_require_message' }),
          type: 'text',
          label: formatMessage({ id: 'app_user_fullName' }),
        }}
        register={register}
        errors={errors}
      />

      <FormControl fullWidth>
        <p className={classes.inputLabel}>
          <FormattedMessage id="app_user_phone" />
        </p>
        <PhoneInput
          defaultCountry="ID"
          value={phone}
          onChange={setPhone}
          className={classes.input}
          data-testid="phone-input"
        />
      </FormControl>

      <InputRHF
        input={{
          name: 'password',
          required: formatMessage({ id: 'app_user_password_require_message' }),
          type: showPass ? 'text' : 'password',
          label: formatMessage({ id: 'app_user_password' }),
          minLength: 8,
          messageMin: formatMessage({ id: 'app_user_password_min_length' }),
        }}
        register={register}
        errors={errors}
      />
      <InputRHF
        input={{
          name: 'confirmPassword',
          required: formatMessage({ id: 'app_user_confirm_password_require_message' }),
          type: showPass ? 'text' : 'password',
          label: formatMessage({ id: 'app_user_confirm_password' }),
          minLength: 8,
          messageMin: formatMessage({ id: 'app_user_confirm_password_min_length' }),
        }}
        register={register}
        errors={errors}
      >
        <label htmlFor="show" className={classes.showPassword}>
          <input
            type="checkbox"
            name="show"
            id="show"
            onChange={(e) => setShowPass(e.target.checked)}
            data-testid="button-show-pass"
          />
          <FormattedMessage id="app_user_password_show" />
        </label>
      </InputRHF>
      <div className={classes.buttonWrap}>
        <button
          type="button"
          className={classes.buttonSubmit}
          onClick={() => dispatch(actionSetStep(step - 1))}
          data-testid="button-back"
        >
          <FormattedMessage id="app_back" />
        </button>
        <button
          type="submit"
          className={classes.buttonSubmit}
          disabled={!fullName || !password || !phone}
          data-testid="button-submit"
        >
          <FormattedMessage id="app_header_register" />
        </button>
      </div>
    </form>
  );
};

RegisterForm.propTypes = {
  intl: PropTypes.object,
  step: PropTypes.number,
  email: PropTypes.string,
  role: PropTypes.string,
};

const mapStateToProps = createStructuredSelector({
  step: selectStep,
  email: selectEmail,
  role: selectRole,
});

export default injectIntl(connect(mapStateToProps)(RegisterForm));
