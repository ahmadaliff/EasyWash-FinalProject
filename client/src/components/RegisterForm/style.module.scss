@import '@styles/base';
.showPassword,
.nav {
  @include directionX;
  align-items: center;
  padding-top: 0.25rem;
  font-size: small;
  color: var(--color-primary);
}
.nav {
  color: var(--color-text);
  a {
    color: var(--color-primary);
  }
}
.toggleRole {
  button {
    color: var(--color-primary);
    text-transform: capitalize;
    border-color: var(--color-primary);
  }
}
.labelRole {
  margin: 0;
  padding-bottom: 0.25rem;
  font-size: small;
}
.buttonWrap {
  @include directionX;
  justify-content: space-between;
  gap: 0.5rem;
  .buttonSubmit {
    width: 100%;
    background-color: var(--color-primary);
    color: var(--color-bg);
    border-radius: 2.5rem;
    padding: 0.5rem 2.5rem 0.5rem 2.5rem;
    margin: auto;
    margin-top: 0.75rem;
    box-shadow: none;
    border: none;
    cursor: pointer;
    &:hover {
      transform: scale(1.0125, 1.0125);
    }
    &:disabled {
      background-color: rgba(black, 0.2);
      color: rgba($color: black, $alpha: 0.5);
    }
    &:disabled:hover {
      transform: none;
    }
  }
}
