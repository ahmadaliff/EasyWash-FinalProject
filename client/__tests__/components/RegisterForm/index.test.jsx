import RegisterForm from '@components/RegisterForm';
import { act, fireEvent, render, screen } from '@utils/testHelper';

let wrapper;
let inputFullName;
let inputPhone;
let inputPassword;
let ConfirmPassword;

const mockProps = {
  intl: { formatMessage: jest.fn() },
};
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

beforeEach(() => {
  act(() => {
    wrapper = render(<RegisterForm {...mockProps} />);
  });
});
describe('RegisterForm Component', () => {
  test('Correct render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('form-register')).toBeInTheDocument();
  });
  test('Should input properly', () => {
    const { getByTestId } = wrapper;
    const buttonShowPass = getByTestId('button-show-pass');
    inputFullName = screen.getByLabelText('Nama Lengkap');
    inputPhone = getByTestId('phone-input');
    inputPassword = screen.getByLabelText('Kata Sandi');
    ConfirmPassword = screen.getByLabelText('Konfirmasi Kata Sandi');
    fireEvent.change(inputFullName, { target: { value: 'Ahmad Alif Sofian' } });
    fireEvent.change(inputPhone, { target: { value: '+6281302123131' } });
    fireEvent.change(inputPassword, { target: { value: 'password123' } });
    fireEvent.change(ConfirmPassword, { target: { value: 'password123' } });
    expect(inputFullName.value).toBe('Ahmad Alif Sofian');
    expect(inputPhone.value).toBe('+62 813 0212 3131');
    expect(inputPassword.value).toBe('password123');
    expect(ConfirmPassword.value).toBe('password123');
    fireEvent.click(buttonShowPass);
    expect(inputPassword.type).toBe('text');
  });
  test('Should call function onsubmit properly', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-submit');
    act(() => {
      fireEvent.click(button);
    });
  });
  test('Should call click button properly', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-back');
    fireEvent.click(button);
  });

  test('Should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
