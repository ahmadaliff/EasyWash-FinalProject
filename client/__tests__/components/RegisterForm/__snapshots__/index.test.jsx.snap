// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm Component Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        data-testid="form-register"
      >
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
          data-testid="inputRHF"
        >
          <label
            class="inputLabel"
            for="fullName"
          >
            Nama Lengkap
            <input
              class="input undefined"
              id="fullName"
              name="fullName"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
          data-testid="inputRHF"
        >
          <label
            class="inputLabel"
            for="phone"
          >
            Telp/No.Hp
            <input
              class="input undefined"
              id="phone"
              name="phone"
              type="number"
              value=""
            />
          </label>
        </div>
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
          data-testid="inputRHF"
        >
          <label
            class="inputLabel"
            for="password"
          >
            Kata Sandi
            <input
              class="input undefined"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </label>
        </div>
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
          data-testid="inputRHF"
        >
          <label
            class="inputLabel"
            for="confirmPassword"
          >
            Konfirmasi Kata Sandi
            <input
              class="input undefined"
              id="confirmPassword"
              name="confirmPassword"
              type="password"
              value=""
            />
          </label>
          <label
            class="showPassword"
            for="show"
          >
            <input
              data-testid="button-show-pass"
              id="show"
              name="show"
              type="checkbox"
            />
            Tampilkan Kata Sandi
          </label>
        </div>
        <div
          class="buttonWrap"
        >
          <button
            class="buttonSubmit"
            data-testid="button-back"
            type="button"
          >
            Kembali
          </button>
          <button
            class="buttonSubmit"
            data-testid="button-submit"
            disabled=""
            type="submit"
          >
            Daftar
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      data-testid="form-register"
    >
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        data-testid="inputRHF"
      >
        <label
          class="inputLabel"
          for="fullName"
        >
          Nama Lengkap
          <input
            class="input undefined"
            id="fullName"
            name="fullName"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        data-testid="inputRHF"
      >
        <label
          class="inputLabel"
          for="phone"
        >
          Telp/No.Hp
          <input
            class="input undefined"
            id="phone"
            name="phone"
            type="number"
            value=""
          />
        </label>
      </div>
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        data-testid="inputRHF"
      >
        <label
          class="inputLabel"
          for="password"
        >
          Kata Sandi
          <input
            class="input undefined"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        data-testid="inputRHF"
      >
        <label
          class="inputLabel"
          for="confirmPassword"
        >
          Konfirmasi Kata Sandi
          <input
            class="input undefined"
            id="confirmPassword"
            name="confirmPassword"
            type="password"
            value=""
          />
        </label>
        <label
          class="showPassword"
          for="show"
        >
          <input
            data-testid="button-show-pass"
            id="show"
            name="show"
            type="checkbox"
          />
          Tampilkan Kata Sandi
        </label>
      </div>
      <div
        class="buttonWrap"
      >
        <button
          class="buttonSubmit"
          data-testid="button-back"
          type="button"
        >
          Kembali
        </button>
        <button
          class="buttonSubmit"
          data-testid="button-submit"
          disabled=""
          type="submit"
        >
          Daftar
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
